oct/tournament/views.py

```diff
@@ -36,6 +41,68 @@
 def error_500(request):
     return render(request, "tournament/error_500.html")
 
+def generate_roles_dict(roles, testing=False):
+    role_list = list(map(lambda r: (r.name[0]+r.name[1:]).lower(), roles))
+
+    if testing == True:
+        return dict(map(lambda role: (role, True), ["host", "registered_player", "custom_mapper", "mappooler", "playtester", "streamer", "commentator", "referee"]))
+    return dict(map(lambda role: (role, role in role_list), ["host", "registered_player", "custom_mapper", "mappooler", "playtester", "streamer", "commentator", "referee"]))
+
+# Login endpoints
+
+@require_POST
+def login_view(request):
+    data = json.loads(request.body)
+
+    try:
+        code = data.get('code')
+        if code is not None:
+            user = User.objects.create_user(code)
+            if user is None:
+                return JsonResponse({"status": "error", "message": "Failed to login"})
+            
+            involvement = user.get_tournament_involvement(tournament_iteration=OCT5).first()
+            roles = sorted(involvement.roles.get_roles(), key=lambda r: USER_DISPLAY_ORDER.index(r)) \
+                if involvement is not None else None
+            sorted_roles = generate_roles_dict(roles, True) # erm make sure to set this back to false
+            
+            request.session["user"] = {
+                "osu_id": user.osu_id,
+                "osu_username": user.osu_username,
+                "osu_avatar": user.osu_avatar,
+                "osu_cover": user.osu_cover,
+                "roles": sorted_roles,
+            }
+
+            _login(request, user, backend=settings.AUTH_BACKEND)
+            return JsonResponse({"status": "success"})
+        
+    except requests.HTTPError as exc:
+        print(exc.response.text)
+        log_err(request, exc)
+        return HttpResponseBadRequest()
+    except Exception as exc:
+        print(exc)
+        log_err(request, exc)
+    return HttpResponseServerError()
+
+
+def logout_view(request):
+    if not request.user.is_authenticated:
+        return JsonResponse({'detail': 'You\'re not logged in.'}, status=400)
+
+    _logout(request)
+    request.session["user"] = None
+    return JsonResponse({'detail': 'Successfully logged out.'})
+
+
+@ensure_csrf_cookie
+def session_view(request):
+    if not request.user.is_authenticated:
+        request.session["user"] = None
+        return JsonResponse({'isAuthenticated': False, 'user': request.session.get("user")})
+
+    return JsonResponse({'isAuthenticated': True, 'user': request.session.get("user")})
```

Comment:
All this request.session stuff seems unnecessary? Just let django handle user sessions and create an endpoint that returns a serialized user object of the current user or in every html response include a tag with the data.

Reply:
Similar yeah, but just unnecessary when instead you could have a couple lines serializing the user object. Plus that info should be freshly obtained when a user opens a page, and not just when they login. If you want to add extra stuff, add that functionality to the serializer.

Reply:
I understand, I changed it to get the user object instead. Unless theres a specific reason not to though I'm going to leave the function stuff named as session so I don't have to change everything else in the frontend as well

Reply:
yeah session for the name is fine

Reply:
alright, sounds good